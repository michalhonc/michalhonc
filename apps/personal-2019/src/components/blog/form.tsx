import React, { useState, useCallback, FormEvent } from "react";
// @ts-ignore 7016
import addToMailchimp from "gatsby-plugin-mailchimp";
import { useIntl } from "gatsby-plugin-intl";

import EnvelopeSvg from "../../images/envelope.svg";
import LoaderSvg from "../../images/loader.svg";
import "../../styles/form.scss";

const IS_ALREADY_REGISTERED = "is already subscribed to list";
const IS_NOT_VALID = "The email you entered is not valid";
const IS_NOT_REAL = "This email address looks fake or invalid";

interface IProps {
  category: string;
  language: string;
  source: string;
}

export const Form: React.FC<IProps> = ({ category, language, source }) => {
  const intl = useIntl();
  const [email, setEmail] = useState("");
  const [firstname, setFirstname] = useState("");
  const [state, setState] = useState("idle");
  const [msg, setMsg] = useState("");

  const handleSubmit = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      setState("loading");
      setMsg("");
      const listFields = {
        FNAME: firstname,
        CATEGORY: category,
        LANG: language,
        SOURCE: source,
      };
      const result = await addToMailchimp(email, listFields);

      setState(result.result);
      setMsg(translateErrorMessage(result.msg, intl));
    },
    [firstname, category, language, source]
  );

  return (
    <section className="Form">
      <div className="Form-left">
        <EnvelopeSvg />
      </div>
      <div className="Form-right">
        <h2>{intl.formatMessage({ id: "form_title" })}</h2>
        <p>{intl.formatMessage({ id: "form_desc" })}</p>
        <form data-state={state} onSubmit={handleSubmit}>
          <input
            type="text"
            onChange={(e) => setFirstname(e.target.value)}
            placeholder={intl.formatMessage({ id: "form_firstname" })}
            required
          />
          <input
            type="email"
            onChange={(e) => {
              setState("idle");
              setMsg("");
              setEmail(e.target.value);
            }}
            placeholder={intl.formatMessage({ id: "form_email" })}
            required
          />
          <div className="Form-message">
            {state === "error" && msg}
            {state === "loading" && <LoaderSvg />}
          </div>
          <input
            type="submit"
            value={intl.formatMessage({ id: "form_submit" })}
          />
        </form>
      </div>
    </section>
  );
};

function translateErrorMessage(msg: string, intl: any) {
  if (msg.includes(IS_NOT_VALID) || msg.includes(IS_NOT_REAL)) {
    return intl.formatMessage({ id: "form_error_isNotValid" });
  }
  if (msg.includes(IS_ALREADY_REGISTERED)) {
    return intl.formatMessage({ id: "form_error_isAlreadyRegistered" });
  }
  return intl.formatMessage({ id: "form_error_general" });
}
