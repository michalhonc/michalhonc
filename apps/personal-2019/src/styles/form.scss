@import 'mixins';

.Form {
  display: flex;
  border-radius: 1rem;
  border: 0.5rem solid var(--c-alpha-light);
  margin: 5rem -2rem 0;

  @include respond('phone') {
    border-radius: 0;
  }

  &-left {
    display: flex;
    justify-content: center;
    align-items: center;
    flex: 1;
    background: var(--c-alpha-light);
    & svg {
      width: 5rem;
      height: 5rem;
      fill: var(--c-alpha);
    }
  }

  &-message {
    color: #F92672;
    height: 3rem;
    font-size: 1.6rem;
    line-height: 2rem;
    & svg {
      width: 3rem;
      height: 3rem;
    }
  }

  &-right {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    flex: 3;
    padding: 0 2rem 2rem;

    & form {
      display: flex;
      flex-direction: column;
      margin-top: 1rem;

      & > *:not(:last-of-type) {
        margin-bottom: 1rem;
      }

      & input[type="email"],
      & input[type="text"] {
        outline: none;
        padding: 1rem;
        border-radius: 0.4rem;
        border: 1px solid #ccc;

        @include respond('phone') {
          max-width: 22rem;
        }

        &:focus {
          border: 1px solid var(--c-alpha);
        }
      }

      & input[type="submit"] {
        outline: none;
        background: var(--c-alpha);
        color: var(--c-white);
        padding: 1rem;
        border-radius: 0.8rem;
        border: none;
        width: 50%;
        cursor: pointer;
      }

      &[data-state="error"] {
        & input[type="email"] {
          border: 1px solid #F92672;
        }

        & input[type="submit"] {
          cursor: default;
          pointer-events: none;
          background: #c5c5c5;
        }

      }
    }
  }
}
