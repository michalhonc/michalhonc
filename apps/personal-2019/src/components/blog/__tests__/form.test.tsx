import * as React from "react";
import renderer from "react-test-renderer";
import { Form } from "../form";

describe("form", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Form category="category" language="cs" source="source" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
