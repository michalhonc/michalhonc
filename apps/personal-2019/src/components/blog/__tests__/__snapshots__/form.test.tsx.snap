// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form renders correctly 1`] = `
<section
  className="Form"
>
  <div
    className="Form-left"
  >
    <test-file-stub />
  </div>
  <div
    className="Form-right"
  >
    <h2>
      Přidejte se do odběru!
    </h2>
    <p>
      Přidejte se do odběru novinek z blogu
    </p>
    <form
      data-state="idle"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Vaše jméno"
        required={true}
        type="text"
      />
      <input
        onChange={[Function]}
        placeholder="E-mail"
        required={true}
        type="email"
      />
      <div
        className="Form-message"
      />
      <input
        type="submit"
        value="Odeslat"
      />
    </form>
  </div>
</section>
`;
